File: ml-benchmark/synduce/constraints/alist/sums.f
Inductive List = nil Unit | cons {Int, List};
Inductive Nat = z Unit | s Nat;

is_unique =
  let key_differ = \key: Int. fix (
    \f: List -> Bool. \xs: List.
    match xs with
      nil _ -> true
    | cons {h, t} -> and (not (== h key)) (f t)
    end
  ) in fix (
    \f: List -> Bool. \xs: List.
    match xs with
      nil _ -> true
    | cons {h, t} -> and (key_differ h t) (f t)
    end
  );

@Input w: Int;

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> if == h w then + h (f t) else f t
  end
);

target = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    nil _ -> nil unit
  | cons {h, t} -> if == h w then
    cons {h, t} /* Avoid the recusion of t*/
    else cons {h, f t}
  end
);

main = \xs: List. if is_unique xs then spec (target xs) else 0;

File: ml-benchmark/synduce/constraints/alist/sums.ml
@Input val w: int

type list = Nil of unit | Cons of int * list
type nat = Z of unit | S of nat

let is_unique =
  let key_differ key =
    let rec f xs =
      match xs with
      | Nil _ -> true
      | Cons (h, t) -> (not (h == key)) && f t
    in
    f
  in
  let rec f xs =
    match xs with
    | Nil _ -> true
    | Cons (h, t) -> (key_differ h t) && f t
  in
  f

let spec =
  let rec f xs =
    match xs with
    | Nil _ -> 0
    | Cons (h, t) -> if h == w then h + f t else f t
  in
  f

val target: list -> list compress
let target =
  let rec f xs =
    match xs with
    | Nil _ -> Nil ()
    | Cons (h, t) -> if h == w then Cons (h, t) else Cons (h, f t)
  in
  f

let program xs = if is_unique xs then spec (target xs) else 0